orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -958,6 +962,57 @@ IMPLEMENT_GRADIENT_BUILDER(GetGeluGradient) {
               {GI(0)})};
 }
 
+namespace {
+std::vector<NodeDef> GetBiasGeluGradNodes(
+    bool use_approximation,
+    const ArgDef& dY, const ArgDef& X, const ArgDef& B,  // inputs
+    const ArgDef& dX, const ArgDef& dB) {                // outputs
+  const auto B_shape = GetShape(B);
+  ORT_ENFORCE(B_shape.size() == 1, "B must have exactly one dimension.");
+
+  const std::vector<int64_t> B_axes = [&B_shape, &X]() {
+    std::vector<int64_t> result{};
+    ComputeBroadcastBackwardAxes(B_shape, GetShape(X), &result, nullptr);
+    return result;
+  }();
+
+  return std::vector<NodeDef>{
+      NodeDef(OpDef{use_approximation ? "BiasFastGeluGrad_dX" : "BiasGeluGrad_dX", kMSDomain, 1},
+              {dY, X, B},
+              {dX}),
+      NodeDef("ReduceSum",
+              {dX},
+              {dB},
+              {{"keepdims", MakeAttribute("keepdims", int64_t{0})},
+               {"axes", MakeAttribute("axes", B_axes)}})};
+}
+}  // namespace
+
+IMPLEMENT_GRADIENT_BUILDER(GetBiasGeluGradient) {
+  const auto dY = GO(0), X = I(0), B = I(1),
+             dX = GI(0), dB = GI(1);
+  return GetBiasGeluGradNodes(false, dY, X, B, dX, dB);
+}
+
+IMPLEMENT_GRADIENT_BUILDER(GetFastGeluGradient) {
+  const auto dY = GO(0), X = I(0),
+             dX = GI(0);
+  const auto num_src_node_inputs = GetSrcNodeInputSize();
+  if (num_src_node_inputs == 2) {  // with bias
+    // FastGeluGrad doesn't support bias - it needs to be composed with other ops
+    const auto B = I(1),
+               dB = GI(1);
+    return GetBiasGeluGradNodes(true, dY, X, B, dX, dB);
+  }
+  if (num_src_node_inputs == 1) {  // without bias
+    return std::vector<NodeDef>{
+        NodeDef(OpDef{"FastGeluGrad", kMSDomain, 1},
+                {dY, X},
+                {dX})};
+  }
+  ORT_THROW("Unexpected number of FastGelu inputs: ", num_src_node_inputs);
```

Comment:
>FastGelu [](start = 34, length = 8)

nit : FastGeluGrad

Reply:
it's referring to the forward op here, FastGelu

---
In reply to: [445308723](https://github.com/microsoft/onnxruntime/pull/4146#discussion_r445308723) [](ancestors = 445308723)